docs: WIP add naught coin exercise (#211)
crytic/building-secure-contracts: program-analysis/echidna/Exercise-X.md

### What you will learn

By completing this exercise, you will:

- Practice defining good invariants in plain English.
- Understand different ways of creating an external testing setup.
- Learn how to write simple properties and improve them over time.

### Exercise setup

The `NaughtCoin` contract has several dependencies. The directory structure that you should aim for looks like this:

```
src
├── NaughtCoin.sol
├── crytic
│   ├── ExternalTest.sol
│   └── config.yaml
└── level-utils
    ├── Context.sol
    ├── ERC20.sol
    ├── IERC20.sol
    └── IERC20Metadata.sol
```

Depending on your framework, the `NaughtCoin` contract will be in the `src` folder (Foundry) or `contracts` (Hardhat). OpenZeppelin dependencies are in the `level-utils` folder, and all Echidna-related files are in the `crytic` folder.

### Exercise goals

> The main goal is to make the player's balance equal to `0`.  You might know the answer right away. However, the purpose of this exercise is to define the whole system using properties in an iterative way. Start small and see what you can find. Expand over time. Here are some steps that you may take:

Author: ```suggestion
The main goal is to make the player's balance equal to `0`.  You might know the answer right away. However, the purpose of this exercise is to describe the whole system using properties in an iterative way. Start small and see what you can find. Expand over time. Here are some steps that you may take:
```

#### Example invariants in plain English

<details>
  <summary>1st invariant </summary>

   The token should be deployed `(address(token) != address(0))`.

</details>

<details>
  <summary>2nd invariant </summary>

   The player token balance should equal the initial supply if the current `block.timestamp < timelock`.

</details>

<details>
  <summary>3rd invariant </summary>

   The token transfer should fail if the current `block.timestamp < timelock`.

</details>

<details>
  <summary>4th invariant </summary>

   The `approve` function should never fail if the caller has a sufficient token balance.

</details>

<details>
  <summary>5th invariant </summary>

   The player should not be able to burn tokens before the `timelock`.

</details>

<details>
  <summary>6th invariant</summary>

   The token transfer via `transferFrom` should fail if the current `block.timestamp < timelock` and/or spender has enough allowance.

</details>

#### External testing setup

<details>
  <summary>Example state setup</summary>

   ```
   // SPDX-License-Identifier: MIT
  pragma solidity ^0.8.0;
  
  import {NaughtCoin} from "src/NaughtCoin.sol";
  

>   contract ExternalTestSimple {

Author: ```suggestion
  contract ExternalTest {
```

As there is no advanced setup followup, remove the word simple.

   address player;
      NaughtCoin public naughtCoin;
  
      constructor() {
       player = msg.sender;

>           naughtCoin = new NaughtCoin(player);

Author: ```suggestion
      naughtCoin = new NaughtCoin(player);
```

Remove spacing

      }
  }
   ```

</details>

<details>
  <summary>See if Echidna works as expected</summary>

   ```
   function always_true() public pure {
        assert(true);
    }
   ```

</details>

#### Writing properties

<details>
  <summary>1st property</summary>

   ```
   function token_is_deployed() public {
        assert(address(naughtCoin) != address(0));
    }
   ```

</details>

<details>
  <summary>2nd property</summary>

   ```

>    function sender_balance_is_equal_to_initial_supply() public {

Author: ```suggestion
   function player_balance_is_equal_to_initial_supply() public {
```

Change sender to player

        assert(naughtCoin.balanceOf(player) == naughtCoin.INITIAL_SUPPLY());
    }
   ```

</details>

<details>
  <summary>2nd property improved with pre-conditions</summary>
  
  In plain English we have defined this invariant as The player token balance should equal the initial supply if the current `block.timestamp < timelock`. The second part of this sentence specifies exactly when this property should be tested.
  
   ```

>    function sender_balance_is_equal_to_initial_supply() public {

Author: ```suggestion
   function player_balance_is_equal_to_initial_supply() public {
```

Change sender to player

        public
    {
        // pre-conditions
        uint256 playerBalance = naughtCoin.balanceOf(player);
        if (playerBalance >= amount) {
            // actions
            bool success1 = naughtCoin.approve(bob, amount);
            // post-conditions
            assert(success1);
        }
    }
   ```

</details>

<details>
  <summary>5th property</summary>
  
  Player should not `burn` tokens before the `timeLock` period.
  What is token burning?
  
  You can burn tokens by sending them to the `0` address or by sending them to a non existent address. Sending to the `0` address is not possible because of the `0` address checks in the `ERC20` OZ standard. Sending them to any address shouldn't be possible as well because of the `timeLock`.
  
  This property would be invalidated if any of our previous property were invalidated (which seems to not be the case at the moment). Let's leave this property for now and move on.

</details>

<details>
  <summary>6th property</summary>
  
  The token transfer via `transferFrom` should fail if the current `block.timestamp < timelock` and/or spender has enough allowance.
  

>   This property in its base form is the same as the one with `transfer` function.

Author: ```suggestion
  This property in its base form is the same as the one with the `transfer` function.
```

Add article "the"

  
   ```
   function transfer_from_should_fail_before_timelock_period(uint256 amount)
        public
    {
        // pre-conditions
        uint256 playerBalanceBefore = naughtCoin.balanceOf(player);
        uint256 bobBalanceBefore = naughtCoin.balanceOf(bob);
        uint256 currentTime = block.timestamp;
        if (currentTime < naughtCoin.timeLock()) {
            // actions
            bool success1 = naughtCoin.transferFrom(player, bob, amount);
            if (success1) {
                emit AssertionFailed(amount);
            }
            // post-conditions
            assert(
                naughtCoin.balanceOf(player) == playerBalanceBefore &&
                    naughtCoin.balanceOf(bob) == bobBalanceBefore
            );
        }
    }
   ```

Run Echidna and see what you get. Echidna emits the `AssertionFailed` event with the `amount` of `0`. It means that the `transferFrom` succeeded. If you look at the corpus, you will see that our post-conditions were invalidated. Echidna found a way to change player's balance.

The property `playerBalance == InitialSupply` still holds. You need to guide Echidna a little bit more for it to invalidate this property.

> Try to get into the mindset of exploring the system you are auditing. Since, we have discovered something new about the system, a `transferFrom` function does not fail, as we initially expected, the next logical step would be to test this functionality and make sure that it works properly.

Author: ```suggestion
Try to get into the mindset of exploring the system you are auditing. Since we have discovered something new about the system, a `transferFrom` function does not fail, as we initially expected, the next logical step would be to test this functionality and ensure that it works properly.
```

Author: Change wording

  │ Call sequence:                                                                                                       │
  │ 1.increaseAllowance(0xa329c0648769a73afac7f9381e08fb43dbea72,11579208923731619542357098500868790785326998466564056403│
  │   9457584007913129639935) from: 0x0000000000000000000000000000000000030000 Time delay: 531977 seconds Block delay:   │
  │   12066                                                                                                              │
  │ 2.no_free_tokens_in_transfer_from(9) from: 0x0000000000000000000000000000000000010000 Time delay: 490448 seconds     │
  │   Block delay: 3753                                                                                                  │
  │ 3.sender_balance_is_equal_to_initial_supply() from: 0x0000000000000000000000000000000000030000 Time delay: 1 seconds │
  │   Block delay: 42595                                                                                                 │
  │ Event sequence:                                                                                                      │
  │ Panic(1)
 ```

</details>

#### Fixing a bug

> Last step is to fix a bug that you have found. Try to fix it and re-run your properties. Check if they hold.

Author: ```suggestion
The last step is to fix a bug that you have found. Try to fix it and re-run your properties. Check if they hold.
```

Add article "the"

<details>
  <summary>Fixing a bug part 1</summary>

```
function transferFrom(
        address _from,
        address _to,
        uint256 _amount
    ) public override lockTokens returns (bool) {
        super.transferFrom(_from, _to, _amount);
    }
```

If you run Echidna, you will see that the properties still fail. Why is that?

Add a `Player(address player)` event to your contract and emit it in your properties. Look at the call sequence.

</details>

<details>
  <summary>Fixing a bug part 2</summary>
  

> The player's address is `0x0000...3000`. Echidna is making calls from multiple accounts. It was able to increase the allowance of an address `0x000...1000` and make a call to `transferFrom(player, 0x000...1000, amount)`. This is expected!

Author: ```suggestion
The player's address is `0x0000...3000`. Echidna is making calls from multiple accounts. It was able to increase the allowance of an address `0x000...1000` and make a call to `transferFrom(player, 0x000...1000, amount)`. This is expected!

[This happens because of the `multi-abi`](https://github.com/crytic/building-secure-contracts/blob/master/program-analysis/echidna/using-multi-abi.md). 
```

> The `lockTokens` modifier does not prevent others from making transfers, only the player is constrained.

Author: ```suggestion
The `lockTokens` modifier does not prevent others from making transfers. Only the player is constrained.
```

Split sentences

# Naught Coin - exercise

This exercise is based on the [NaughtCoin - level 15 of the Ethernaut wargame](https://ethernaut.openzeppelin.com/level/15).

The first four exercises were focused on internal testing. This level is an excellent introduction to [external testing with Echidna](https://github.com/crytic/building-secure-contracts/blob/master/program-analysis/echidna/common-testing-approaches.md#external-testing).  Since the `NaughtCoin` contract expects a `player` to be the owner of the NaughtCoin tokens, it is a good idea to go with an external setup. The `EchidnaTest` contract will act as a player. Such a setup allows us to relatively easy describe user2user or user2contract interactions.

### What you will learn

By completing this exercise, you will:

- Practice defining good invariants in plain English.
- Understand different ways of creating an external testing setup.
- Learn how to write simple properties and improve them over time.

### Exercise setup

The `NaughtCoin` contract has several dependencies. The directory structure that you should aim for looks like this:

```
src
├── NaughtCoin.sol
├── crytic
│   ├── ExternalTest.sol

> │   └── config.yaml

Reviewer: Can you expand a bit on what the config should look like?

Author: I will expand on it definitely, along with the expanded setup configuration.

Author: Section added

Reviewer: This is great now. Keep in mind that you can add some links to `using-multi-abi.md` and `testing-modes.md` to avoid some explanations.

Author: Forgot to do so. I am on that!

Author: Links added.

        }
        // post-conditions
        assert(
            naughtCoin.balanceOf(player) == playerBalanceBefore &&
                naughtCoin.balanceOf(bob) == bobBalanceBefore
        );
    }
   ```

Run Echidna and check the corpus. Is this a good property?

</details>

<details>
  <summary>Improving the 3rd property</summary>

   If you look at the corpus:

   ```
 34 | r   |     function transfer_should_fail_before_timelock_period() public {
 35 |     |         // pre-conditions
 36 | r   |         uint256 playerBalanceBefore = naughtCoin.balanceOf(player);
 37 | r   |         uint256 bobBalanceBefore = naughtCoin.balanceOf(bob);
 38 | r   |         uint256 currentTime = block.timestamp;
 39 | r   |         if (currentTime < naughtCoin.timeLock()) {
 40 |     |             // actions
 41 | r   |             naughtCoin.transfer(bob, 100);
 42 |     |         }
 43 |     |         // post-conditions
 44 |     |         assert(
 45 |     |             naughtCoin.balanceOf(player) == playerBalanceBefore &&
 46 |     |                 naughtCoin.balanceOf(bob) == bobBalanceBefore
 47 |     |         );
 48 |     |     }
 49 |     | }
   ```

>    You can see that the transfer reverted as expected. But... Our post-conditions weren't checked. If you have read the [ERC20 spec](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-20.md), you will know that the `transfer` function returns a status `boolean`. In the case of success, we can emit a special `AssertionFailed` event that won't stop the execution of the function. This way, our post-conditions will be checked.

Reviewer: What is your opinion on using `try`/`catch` instead of relying on the return value? OZ tokens either revert or return `true`:
 ```
if (currentTime < naughtCoin.timeLock()) {
         // actions
         try naughtCoin.transfer(bob,amount) { 
              // call succeeded, this is not ok
              assert(false);
         } catch {
              // call reverted
         }
}
```

Author: Thank you so much for your feedback, @glarregay-tob! As I described in my [blog post](https://wizzardhat.com/ethernaut-level-15-naughtcoin/), I initially went for the `try`/`catch`. They are much more intuitive and overall way better than return value checks. I switched to value checks for a couple of reasons:
- I couldn't get Echidna to work with `try`/`catch`. The property was either always passing or always reverting
- That's what Justin did in the [Intro to AMM's invariants workshop](https://www.youtube.com/watch?v=n0RaKKVTGvA), and I thought that it might be a better practice
- I must admit that in this particular case, the `try` / `catch` with an empty catch block, as explained in the [How to write good properties](https://github.com/crytic/building-secure-contracts/blob/master/program-analysis/echidna/property-creation.md) section is a little bit confusing to me. 

I assumed that the OZ standard returns boolean values, but as you pointed out, this is wrong. The `transfer` and `transferFrom` functions return a boolean value, but the internal `_transfer` function reverts on failure. 

Thanks again for pointing that out; I will also fix this in my article. 

I will re-write the properties with `try`/`catch` and update the exercise.

Author: I think it might be helpful to leave this section as it is and create a new follow-up section with Justin's setup using an external `Setup` contract and re-write properties using `try`/`catch` to showcase that both options are valid. What do you think, @glarregay-tob?

Reviewer: It is weird that it didn't work for you. Let's go with the option that works first, and then if you want you can add the new section.
As a side note, this can actually be generic for reverting and non reverting tokens: you can get the return value from a `try` call by using `try contract.function(param) returns (bool success) {` if that helps. 
Maybe "How to write good properties" needs an update too :) thanks for pointing it out!

Author: > As a side note, this can actually be generic for reverting and non reverting tokens: you can get the return value from a try call by using try contract.function(param) returns (bool success) { if that helps. 

This might actually be the case. The `transfer` and `transferFrom` functions return true if they succeed. The reverting mechanism is caused by the internal function `_transfer`. 

I will test it out in the follow-up section. 

Thanks @glarregay-tob!

Author: I've been experimenting with the `try`/`catch`, and this is where I got stuck the last time. 

According to [writing good properties step by step](https://github.com/crytic/building-secure-contracts/blob/master/program-analysis/echidna/property-creation.md#enhacing-postcondition-checks), if we expect a function to revert, we should place the `assert(false)` in the first block. It means that the function call should not succeed, and when it does, we are bringing that up with the `assert(false)`. The following code results in a property violation. 

In this property, we expect that the `transfer` will fail.
```solidity
function token_transfer_should_fail_before_timelock_period(uint256 amount)
        public
    {
        amount = _between(amount, 0, token.INITIAL_SUPPLY());

        // pre-conditions
        uint256 currentTime = block.timestamp;
        uint256 playerBalanceBefore = token.balanceOf(address(player));
        uint256 bobBalanceBefore = token.balanceOf(address(bob));

        if (currentTime < token.timeLock()) {
            // action
            try
                player.proxy(
                    address(token),
                    abi.encodeWithSignature(
                        "transfer(address,uint256)",
                        address(bob),
                        amount
                    )
                )
            {
                assert(false);
            } catch {
                /* reverted */
            }

            // post-conditions
            assert(token.balanceOf(address(player)) == playerBalanceBefore);
            assert(token.balanceOf(address(bob)) == bobBalanceBefore);
        }
    }
```

And this Echidna's event sequence:
```
Call sequence:                                                                                                    
1.token_transfer_should_fail_before_timelock_period(0)                                                            
Event sequence:                                                                                                      
Panic(1)                                                                                                            
merror Revert 0x  
```

The `assert(false)` gets reverted. 

What does the `merror Revert 0x` mean?

I got this property working by applying your suggestion about adding the `returns(bool success)`:

```solidity
try
  player.proxy(
      address(token),
      abi.encodeWithSignature(
          "transfer(address,uint256)",
          address(bob),
          amount
      )
  )
returns (bool success, bytes memory returnData) {
  assert(!success);
} catch {
  /* reverted */
}
```

This, however, leaves me with an unused second return parameter, `returnData`. Should I ignore this warning, or is there a workaround? 

The compiler won't let me ignore the second returned data type. The `returns(bool success)` without the `bytes memory returnData` results in: `TypeError: Function returns 2 values, but returns clause has 1 variables.` 

Emitting this event also does not make sense as it won't be seen by Echidna in any case. 
```solidity
event Log(bytes data);

// ...
returns (bool success, bytes memory returnData) {
    emit Log(returnData);
    assert(!success);
} catch {
    /* reverted */
}
```

I might be missing something obvious.

Reviewer 2: > What does the merror Revert 0x mean?

I believe this is caused by calling an external contract with the 0x0 address. Are you sure all the addresses are initialized?

Author: Hi @ggrieco-tob, Thanks for your response! I took a week off for a winter break. I will test this out as soon as I get back. Thanks 🎉

- If `transfer` failed, that's okay. Keep going.
- If `transfer` succeeded, we don't want that, emit the `AssertionFailed` and check post-conditions as usual.
 
 Add the `event AssertionFailed(uint256 amount);` at the top of your contract.
 
   ```
   function transfer_should_fail_before_timelock_period(uint256 amount)
        public
    {
        // pre-conditions
        uint256 playerBalanceBefore = naughtCoin.balanceOf(player);
        uint256 bobBalanceBefore = naughtCoin.balanceOf(bob);
        uint256 currentTime = block.timestamp;
        if (currentTime < naughtCoin.timeLock()) {
            // actions
            bool success1 = naughtCoin.transfer(bob, amount);
            if (success1) {
                emit AssertionFailed(amount);
            }
        }
        // post-conditions
        assert(
            naughtCoin.balanceOf(player) == playerBalanceBefore &&
                naughtCoin.balanceOf(bob) == bobBalanceBefore
        );
    }

   ```

Run Echidna and check the corpus. Now we have fully covered this property.

</details>

<details>
  <summary>4th property</summary>
  

>   According to the ERC20 spec, the `approve` function should not fail if the caller has enough tokens to make the approval.

Reviewer: `approve` won't fail even if you don't have enough tokens. It's like signing a blank check: "I allow this account to spend up to xxx of my funds in one or more transactions"
It might happen that I don't have the funds yet, but the approval goes through anyway.

Author: This is an oversight on my part. It's called an "allowance overrun", right? 

Will fix that. Thanks!

Author: Does it mean that this property is not the best one to test? If technically it will never fail?

Author: Fixed

Reviewer: I'm not aware of it having a special name, the ERC20 standard does not mention it. You just allow the account to spend your funds.
Reviewing OZ's implementation, the only way it can fail is if the address zero is involved, otherwise it just updates the allowance.

Author: Fixed

  
   ```
   function approve_should_not_fail_if_caller_has_enough_tokens(uint256 amount)
        public
    {
        // pre-conditions
        uint256 playerBalance = naughtCoin.balanceOf(player);
        if (playerBalance >= amount) {
            // actions
            bool success1 = naughtCoin.approve(bob, amount);
            // post-conditions
            assert(success1);
        }
    }
   ```

</details>

<details>
  <summary>5th property</summary>
  
  Player should not `burn` tokens before the `timeLock` period.
  What is token burning?
  

>   You can burn tokens by sending them to the `0` address or by sending them to a non existent address. Sending to the `0` address is not possible because of the `0` address checks in the `ERC20` OZ standard. Sending them to any address shouldn't be possible as well because of the `timeLock`.

Reviewer: This is a bit of a technicality, but transferring it to anywhere doesn't burn them, it just locks them.
A "proper" burning should decrease the total supply of the token, and not all tokens implement public functions to do so. Base [ERC20 OZ token](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/a70ee4e3bbee565e49217c9ec59de5354f8af9d8/contracts/token/ERC20/ERC20.sol#L277-L288) implements an internal _burn method that can be used by token implementations.

Author: This is a fair point. I think the word "effectively" is missing. Sending tokens to a non-existing address locks them up, which effectively burns them. 

The `ERC20` used in this Ethernaut challenge did not implement a public burn functionality.

Author: Fixed

  
  This property would be invalidated if any of our previous property were invalidated (which seems to not be the case at the moment). Let's leave this property for now and move on.

</details>

<details>
  <summary>6th property</summary>
  
  The token transfer via `transferFrom` should fail if the current `block.timestamp < timelock` and/or spender has enough allowance.
  

>   This property in its base form is the same as the one with the `transfer` function.

Reviewer: If the previous one is changed to `try`/`catch`, this one should be changed too

  │   Block delay: 3753                                                                                                  │
  │ 3.sender_balance_is_equal_to_initial_supply() from: 0x0000000000000000000000000000000000030000 Time delay: 1 seconds │
  │   Block delay: 42595                                                                                                 │
  │ Event sequence:                                                                                                      │
  │ Panic(1)
 ```

</details>

#### Fixing a bug

The last step is to fix a bug that you have found. Try to fix it and re-run your properties. Check if they hold.

<details>
  <summary>Fixing a bug part 1</summary>

```
function transferFrom(
        address _from,
        address _to,
        uint256 _amount
    ) public override lockTokens returns (bool) {
        super.transferFrom(_from, _to, _amount);
    }
```

If you run Echidna, you will see that the properties still fail. Why is that?

Add a `Player(address player)` event to your contract and emit it in your properties. Look at the call sequence.

</details>

<details>
  <summary>Fixing a bug part 2</summary>
  

> The player's address is `0x0000...3000`. Echidna is making calls from multiple accounts. It was able to increase the allowance of an address `0x000...1000` and make a call to `transferFrom(player, 0x000...1000, amount)`. This is expected!

Reviewer: Visual tip: We usually write `0x30000` or `0x10000` to avoid the trailing zeros.

Author: Great tip, will fix it. Thanks!

Author: Fixed
